sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -427,6 +433,36 @@ joint_matrix_apply(sg, C, [=](T &x) {
 });
 ```
 
+===== Input and output matrix are different
```

Comment:
The intention of this new overload is to only modify the destination matrix. I will add more clarification to the spec.
I will also add const modifier in the example to emphasize that.

Reply:
I think this is an improvement. However I think it would be better to have an implementation that removed all restrictions that the interface imposes with respect to the cuda runtime implementation. The cuda implementation allows doing operations with an arbitrary number of matrices, provided that they are the same size (because then the mapping is the same to matrix index).

To explicit, this change allows 

```
Matrix_1{nm} = f1(matrix_2{nm}) + f2(matrix_2{nm})
```
where n,m specify a matrix element, and f1/f2 are arbitrary element wise functions.
But what if you want operations like this but for 3 or 4 or 5 matrices?
You can't do that here but you can in CUDA.

I think I remember that the original CUDA-like way of doing it was changed for some intel hardware reason, and also that it is nice to have a lambda. But if you want to keep it like this maybe there can at least be a parameter pack in the lambda that allows an arbitrary number of matrices?
Even for this it might be wise to think carefully about if there are any further restrictions such an interface imposes on programmability.

I still think it would be best to have the `get_wi_marray()` style function that is equivalent to cuda impl and imposes no additional restrictions on programmability.

Reply:
In implementation of apply, we still use `get_wi_data` so not sure what are the reasons to limit interface to `apply` style.

Reply:
The main reason for removing the original way of element-wise operations, which I call the explicit loop style via Intel's get_wi_data and CUDA'wi_marray, is that backends are forced to make per WI distribution while there are other distributions that may be more efficient depending on the target hardware like per row distribution. 
It is true that today we use the loop style in the implementation but we plan on moving away from that as we are introducing a new SPIRV function for the apply. 

Other reasons are: this WI reasoning is hardware specific and hard to reason about (conversion from SG to WI domain) from user perspective. Functionality-wise for the user, the explicit loop style does not add more compared to the apply style as in both cases, operations are performed on all elements of the matrix so there is no benefit of exposing the WI portion.

I agree with Jack that limiting the apply signature made it that you can do things with the explicit loop style that you cannot do with apply today, like this one case. But I am fixing it in this PR.
Our approach has been that we add these cases when they come up :) 

The specific case that came up from CUDA code is:
```
for (int i = 0; i < c_frag.num_elements; i++) {
 c_frag.x[i] = alpha * acc_frag.x[i] + beta * c_frag.x[i];
}
```
with this PR user can write this in SYCL as follows:
```
sycl::ext::oneapi::experimental::matrix::joint_matrix_apply(item.get_sub_group(), acc_frag, c_frag, 
                                    [=](float &x, float &y) { y = alpha * x + beta * y; });  
```
If there is more than one function, I don't think that is an issue.
But if there is more than two matrices, like 
`jm3 = jm1*alpha + jm2*beta;`
, users will have to break it up into two apply functions compared to one line of code with the explicit loop style: `jm3_wi_data[i] =  jm1.wi_data*alpha + jm2.wi_data*beta; `
But at the same time, users do not need to know about size of `wi_data` per WI. 
Performance-wise,  backends that use the loops style underneath will generate multiple loops which may end up fused or not. But in both cases, there should not be performance degradation. 
Having said that, I am open to adding the variadic matrix argument to avoid writing multiple apply functions :)

Reply:
OK sure.